reference-implementation/lib/writable-stream.js

```diff
@@ -4,225 +4,661 @@ const { InvokeOrNoop, PromiseInvokeOrNoop, PromiseInvokeOrFallbackOrNoop, Valida
+
+    if (state === 'waiting') {
+      WritableStreamDefaultWriterReadyPromiseResolve(writer, undefined);
     }
+  }
+
+  stream._state = 'errored';
+  stream._storedError = error;
 
-    ErrorWritableStream(this, reason);
-    const sinkAbortPromise = PromiseInvokeOrFallbackOrNoop(this._underlyingSink, 'abort', [reason], 'close', []);
-    return sinkAbortPromise.then(() => undefined);
+  return WritableStreamDefaultControllerAbort(stream._writableStreamController, reason);
+}
+
+// WritableStream API exposed for controllers.
+
+function WritableStreamAddWriteRequest(stream) {
+  const writer = stream._writer;
+  assert(IsWritableStreamDefaultWriter(writer) === true);
+
+  const state = stream._state;
+  assert(state === 'writable' || state === 'waiting');
+
+  const promise = new Promise((resolve, reject) => {
+    const writeRequest = {
+      _resolve: resolve,
+      _reject: reject
+    };
+
+    stream._writeRequests.push(writeRequest);
+  });
+
+  return promise;
+}
+
+function WritableStreamError(stream, e) {
+  const state = stream._state;
+  assert(state === 'writable' || state === 'waiting' || state === 'closing');
+
+  const writeRequests = stream._writeRequests;
+  while (writeRequests.length > 0) {
+    const writeRequest = writeRequests.shift();
+    writeRequest._reject(e);
   }
 
-  close() {
-    if (!IsWritableStream(this)) {
-      return Promise.reject(new TypeError('WritableStream.prototype.close can only be used on a WritableStream'));
+  const writer = stream._writer;
+
+  if (writer !== undefined) {
+    WritableStreamDefaultWriterClosedPromiseReject(writer, e);
+
+    if (state === 'waiting') {
+      WritableStreamDefaultWriterReadyPromiseResolve(writer, undefined);
     }
+  }
 
-    if (this._state === 'closing') {
-      return Promise.reject(new TypeError('cannot close an already-closing stream'));
+  stream._state = 'errored';
+  stream._storedError = e;
+}
+
+function WritableStreamFinishClose(stream) {
+  assert(stream._state === 'closing');
+
+  stream._state = 'closed';
+
+  const writer = stream._writer;
+
+  // writer cannot be released while close() is ongoing. So, we can assert that
+  // there's an active writer.
+  assert(writer !== undefined);
+
+  WritableStreamDefaultWriterClosedPromiseResolve(writer);
+}
+
+function WritableStreamFulfillWriteRequest(stream) {
+  assert(stream._writeRequests.length > 0);
+
+  const writeRequest = stream._writeRequests.shift();
+  writeRequest._resolve(undefined);
+}
+
+function WritableStreamUpdateBackpressure(stream, backpressure) {
+  const state = stream._state;
+  const writer = stream._writer;
+
+  if (state === 'writable') {
+    if (backpressure === false) {
+      return;
     }
-    if (this._state === 'closed') {
-      return Promise.reject(new TypeError('cannot close an already-closed stream'));
+
+    stream._state = 'waiting';
+
+    if (writer !== undefined) {
+      WritableStreamDefaultWriterReadyPromiseReset(writer);
```

Comment:
I think this is a helper function, but in the spec language we would not write it as an abstract op. Because in the spec language we don't need to track the resolve and reject callbacks, so it can be a very simple "Set writer.[[readyPromise]] to a new promise." The same might be true of many of the other promise-manipulation methods.

In general naming them with lowerCamelCase and maybe moving them to another block would help differentiate when it comes time to do the spec transition.